src/trio/_repl.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import ast
+import contextlib
+import inspect
+import sys
+import types
+import warnings
+from code import InteractiveConsole
+from typing import Generator
+
+import trio
+import trio.lowlevel
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
+
+
+def _flatten_exception_group(
+    excgroup: BaseExceptionGroup[BaseException],
+) -> Generator[BaseException, None, None]:
+    for exc in excgroup.exceptions:
+        if isinstance(exc, BaseExceptionGroup):
+            yield from _flatten_exception_group(exc)
+        else:
+            yield exc
+
+
+class TrioInteractiveConsole(InteractiveConsole):
+    # code.InteractiveInterpreter defines locals as Mapping[str, Any]
+    # but when we pass this to FunctionType it expects a dict. So
+    # we make the type more specific on our subclass
+    locals: dict[str, object]
+
+    def __init__(self, repl_locals: dict[str, object] | None = None):
+        super().__init__(locals=repl_locals)
+        self.compile.compiler.flags |= ast.PyCF_ALLOW_TOP_LEVEL_AWAIT
+
+    def runcode(self, code: types.CodeType) -> None:
+        async def _runcode_in_trio() -> BaseException | None:
+            func = types.FunctionType(code, self.locals)
+            try:
+                coro = func()
+            except BaseException as e:
+                return e
+
+            if inspect.iscoroutine(coro):
+                try:
+                    await coro
+                except BaseException as e:
+                    return e
+            return None
+
+        maybe_exc_or_excgroup = trio.from_thread.run(_runcode_in_trio)
+
+        if maybe_exc_or_excgroup is not None:
+            # maybe_exc_or_excgroup is an exception, or an exception group.
+            # If it is SystemExit or if the exception group contains
+            # a SystemExit, quit the repl. Otherwise, print the traceback.
+            if isinstance(maybe_exc_or_excgroup, SystemExit):
+                raise maybe_exc_or_excgroup
+            elif isinstance(maybe_exc_or_excgroup, BaseExceptionGroup):
+                sys_exit_exc = maybe_exc_or_excgroup.subgroup(SystemExit)
+                if sys_exit_exc:
+                    # There is a SystemExit exception, but it might be nested
+                    # If there are more than one SystemExit exception in
+                    # the group, this will only find and re-raise the first.
+                    raise next(_flatten_exception_group(sys_exit_exc))
```

Comment:
It might be worth printing all non-system exit exceptions in the exception group?

Reply:
I like this idea. I just need to have a think how to do that in a way that isn't too messy.

Reply:
I think I should remove that elif entirely. Then the exception group will be printed and the repl won't exit. Ideally, there shouldn't be a call to `sys.exit()` buried down the call stack that would cause this, but if there is, it is unlikely the user intended to exit the interpreter.

Trying to get me head around this, there are the following cases...
1. User types `exit()`, intending to quit the repl.
2. Some code called from within the repl calls `sys.exit()` or otherwise raises `SystemExit()`
3. Some code raises an exception group which includes a `SystemExit()`. Perhaps a call to `sys.exit()` inside a task.

For cases 1 & 2, this repl, standard python and asyncio python all just exit with no traceback. So if there is a call to `sys.exit()` buried down the call stack somewhere, the repl just quits. It's not easy to tell the different between 1 & 2 so this is reasonable even if it would be better to show the traceback in the case of 2.

Case 3 is similar to case 2, it's just that there is most likely some async code involved. We don't have the problem case 2 has, having to distinguish from a "real" command to exit the repl. So we stop treating this as a special case, then the code is simpler and the user gets better information.

In the `python -m asyncio` repl, a `SystemExit()` inside a `BaseExceptionGroup` doesn't quit the interpreter. Instead, the exception group gets printed like any other exception. The standard python interpreter is the same. So the proposed change would be in line with the behavior there.

Does that sound reasonable?

Reply:
I've made this change and pushed. Let me know what you think and I can keep it or roll back if needed.

Reply:
It feels a bit weird to me to have this different behavior, but now that I think of it it's probably nicer. If I write `exit()` inside a nursery I probably don't exist my whole repl to exit!